WIP UShER created (DRAFT)

sassy-crick opened the pull request:
(created using `eb --new-pr`)

sassy-crick commented:
I am struck at the `tbb` configuration right at the beginning. Newer versions like the latest from [here](https://github.com/easybuilders/easybuild-easyconfigs/pull/13404) don't seem to have the required `TBBBuild.cmake` file, so I reverted to this one which has it. However, the installation failed here:
```CMake Error at CMakeLists.txt:38 (include):
  include could not find requested file:

    /cmake/TBBBuild.cmake


CMake Error at CMakeLists.txt:39 (tbb_build):
  Unknown CMake command "tbb_build".
```

That line in `CMakeLists.txt` reads:
```include(${TBB_DIR}/cmake/TBBBuild.cmake)```

Even if I change module file to include 
```setenv("TBB_DIR", "/home/sassy/apps/software/tbb/2020.3-GCCcore-10.3.0/")```
I am still stuck with the same problem. 

Also, loading all the required modules and trying to compile it manually, I get this error:

```-- Intel TBB can not be built: Makefile or src directory was not found in 
CMake Error at CMakeLists.txt:41 (find_package):
  By not providing "FindTBB.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "TBB", but
  CMake did not find one.

  Could not find a package configuration file provided by "TBB" with any of
  the following names:

    TBBConfig.cmake
    tbb-config.cmake

  Add the installation prefix of "TBB" to CMAKE_PREFIX_PATH or set "TBB_DIR"
  to a directory containing one of the above files.  If "TBB" provides a
  separate development package or SDK, be sure it has been installed.```

when I am hard-coding the `TBB_DIR` path.

sassy-crick commented:
Downgraded the toolchain to `gompi-2020a` as `tbb-2021.3.0` [PR 13404](https://github.com/easybuilders/easybuild-easyconfigs/pull/13404) is not ready yet: `tbb/pipeline.h` is missing, apparently. 
This downgrade means I will update [PR 13119](https://github.com/easybuilders/easybuild-easyconfigs/pull/13119).
The `boost` library is already merged in the `development` branch.

sassy-crick commented:
What I have not included is this (taken from [here](https://github.com/yatisht/usher/blob/master/installUbuntu.sh)
# install faToVcf
rsync -aP rsync://hgdownload.soe.ucsc.edu/genome/admin/exe/linux.x86_64/faToVcf .
chmod +x faToVcf
mv faToVcf build/

sassy-crick commented:
Following the suggestion of @branfosj  from PR 13708 I have added `Kent_tools` which includes `faToVcf` .
It does make sense to have all of that in one EC instead of loading it separately.

SethosII reviewed (changes requested):
@sassy-crick I would split the patch into two patches because they address different things that aren't connected. Also the usual name for patches is like `%(name)s-%(version)s_fix-something.patch` so in your case it's something like `UShER-0.4.1_fix-isnan.patch` and `UShER-0.4.1_disable-internal-tbb.patch` or so.

sassy-crick commented:
@SethosII I renamed the patch, that slipped through. I also had some styled problems which oddly enough slipped through as well. 
As for splitting the patch: I think I leave that up to a maintainer to decided. I don't have an opinion on it.

Micket commented on easybuild/easyconfigs/u/UShER/UShER-0.4.1-gompi-2020b.eb:
```diff
@@ -0,0 +1,45 @@
+    ('Boost', '1.74.0'),
+    ('Python', '3.8.6'),
+    ('zlib', '1.2.11'),
+    ('Kent_tools', '411'),
```
This doesn't depend on Kent tools at all. It only uses one specific tool in one example workflow file, which is a file that we don't even install. So there is literally not a single reference to this software anywhere.

sassy-crick replied:
I beg to differ. The installation [instructions](https://github.com/yatisht/usher/blob/master/installCentOS.sh) clearly include faToVcf which is part of the the Kent_tools set. Anybody who is using `UShER` from the GitHub page thus expects this program to be around when installing UShER via EasyBuild. So for me this is an extension similarly to the modules we are installing in `R`. So from that angle I think it makes much sense to install that. This is particularly true as ''UShER'' is part of the ''pagnolin'' package which is widely used for Covid-19 PCR tests. It does not help to exclude this vital program, which is needed to convert one file format into another one. 
I agree with the part of how they included it is not the best way, but I disagree on the part regarding dependency. In order to use the full potential of the program, and in order how it is used in praxis, we should leave the `Kent_tool` in it as a dependency.

Micket replied:
It is only referenced in
1. A snakefile that we don't even install. https://github.com/yatisht/usher/blob/master/workflows/Snakefile#L82
2. A snakefile that requires (according to their own instructions) to use conda, which, of course won't be compatible with EB regardless.
Nothing that gets installed here has any connection to the other tools whatsoever.

Micket commented on easybuild/easyconfigs/u/UShER/usher-0.4.1-isnan-CMake.patch:
```diff
@@ -0,0 +1,43 @@
+Patch to correct CMakeLists.txt so external TBB will be used instead of internal build.
+Patch to convert innan to std::isnan
```
Please do these in separate patches. Also, you don't actually need to change the std::isnan for 10.3.0 at least. 
Also the rest of the changes are unncessary. Using find_package works fine. cf. #13708

sassy-crick replied:
I am happy to split the patch in two, no worries here. 
I cannot use 10.3.0, simply because of the problem I am having with `tbb-2021.3.0` and the missing header file so I am stuck with the older version where we only have the 10.2.0 version of `gcc`.
I am not sure where the std::isnan crept in. I cannot reproduce it any more so I removed that patch, only leaving the patch for CMake. 
I had problems with the CMake patch when I tried that method manually which caused the long discussion thread. 
The patch I came up with works also outside EasyBuild, which I see as an advantage specially when debugging programs. In the end, both work.

sassy-crick replied:
I have removed the usher-0.4.1-isnan-CMake.patch and replaced it with the usher-0.4.1-CMake.patch one which addresses the CMakeList.txt issue we both noted, and solved differently.

sassy-crick commented:
To move forward and get that in dry sheets: I have removed the Kent_tools from this PR and added a line both into the module comment and in the EB file. I suggest we merge this PR and [PR 13708](https://github.com/easybuilders/easybuild-easyconfigs/pull/13708). I did not want to 'steal' the patch file from the other PR. 
@Micket are you happy with this merge?
